ops/log.py

```diff
@@ -89,5 +93,234 @@ def except_hook(
+    """Security event names for sequence events.
+
+    See https://cheatsheetseries.owasp.org/cheatsheets/Logging_Vocabulary_Cheat_Sheet.html
+    """
+
+    SEQUENCE_FAIL = 'sequence_fail'
+
+
+class _SecurityEventSession(enum.Enum):
+    """Security event names for session events.
+
+    See https://cheatsheetseries.owasp.org/cheatsheets/Logging_Vocabulary_Cheat_Sheet.html
+    """
+
+    SESSION_CREATED = 'session_created'
+    SESSION_RENEWED = 'session_renewed'
+    SESSION_EXPIRED = 'session_expired'
+    SESSION_USE_AFTER_EXPIRE = 'session_use_after_expire'
+
+
+class _SecurityEventSystem(enum.Enum):
+    """Security event names for system events.
+
+    See https://cheatsheetseries.owasp.org/cheatsheets/Logging_Vocabulary_Cheat_Sheet.html
+    """
+
+    SYS_STARTUP = 'sys_startup'
+    SYS_SHUTDOWN = 'sys_shutdown'
+    SYS_RESTART = 'sys_restart'
+    SYS_CRASH = 'sys_crash'
+    SYS_MONITOR_DISABLED = 'sys_monitor_disabled'
+    SYS_MONITOR_ENABLED = 'sys_monitor_enabled'
+
+
+class _SecurityEventUser(enum.Enum):
+    """Security event names for user events.
+
+    See https://cheatsheetseries.owasp.org/cheatsheets/Logging_Vocabulary_Cheat_Sheet.html
+    """
+
+    USER_CREATED = 'user_created'
+    USER_UPDATED = 'user_updated'
+    USER_ARCHIVED = 'user_archived'
+    USER_DELETED = 'user_deleted'
+
+
+_SecurityEvent = typing.Union[
+    _SecurityEventAuthN,
+    _SecurityEventAuthZ,
+    _SecurityEventCrypt,
+    _SecurityEventExcess,
+    _SecurityEventUpload,
+    _SecurityEventInput,
+    _SecurityEventMalicious,
+    _SecurityEventPrivilege,
+    _SecurityEventSensitive,
+    _SecurityEventSequence,
+    _SecurityEventSession,
+    _SecurityEventSystem,
+    _SecurityEventUser,
+]
+
+
+def _log_security_event(
+    # These are the OWASP log levels, which are not the same as the Juju or Python log levels.
+    level: typing.Literal['INFO', 'WARN', 'CRITICAL'],
+    event_type: _SecurityEvent | str,
+    event: str,
+    *,
+    description: str,
+):
+    """Send a structured security event log to Juju, as defined by SEC0045.
+
+    Args:
+        level: log level of the security event (this is not the same as the Juju log level)
+        event_type: the event type, in the format described by OWASP
+          https://cheatsheetseries.owasp.org/cheatsheets/Logging_Vocabulary
+        event: the name of the event, in the format described by OWASP
+        description: a free-form description of the event, meant for human
+          consumption. Includes additional details of the event that do not
+          fit in the event name.
+    """
+    logger = logging.getLogger(__name__)
+    for juju_handler in logger.handlers:
+        if isinstance(juju_handler, JujuLogHandler):
+            model_backend = juju_handler.model_backend
+            juju_context = model_backend._juju_context
+            app_id = f'{juju_context.model_uuid}-{juju_context.unit_name}'
+            juju_log = model_backend.juju_log
```

Comment:
This avoids having to get things from the environment and means we can directly use the hook command rather than go through the logger, which is the better approach I think.

It is responsible for most of the failing tests - the tests skip setting up the logging (charms can't do that if they are going to use the framework) so fall through to the else block. If this appropriate is acceptable then I can adjust the tests.